---
title: "CENP ChiP-seq ideogram"
author: "Izaskun Mallona at Mark D. Robinson lab, University of Zurich"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: 
  html_document:
    toc: true
    toc_float: true
    code_folding: hide
    code_download: true
    number_sections: true
    df_print: kable
    theme: lumen
params:
  seed: 6
---

```{r}
.libPaths('/home/imallona/R/R4_bioc314/')
```

```{r, message = FALSE}
suppressPackageStartupMessages({
    library(ChIPseeker)
    library(knitr)
    library(GenomicRanges)
    ## library(genomation)
    ## library(Rsamtools)
    ## library('TxDb.Mmusculus.UCSC.mm10.knownGene')
    ## library(AnnotationHub)
    ## library(rtracklayer)
    library(org.Mm.eg.db)
})
```




```{r, echo = FALSE, warning = FALSE, message = FALSE}
knitr::opts_chunk$set(autodep = TRUE, cache = FALSE,
                      cache.lazy = FALSE,
                      dev = "png",
                      fig.width = 5, fig.height = 5,
                      dev.args = list(png = list(type = "cairo")),
                      warning = FALSE, message = FALSE)
```

```{r}
DATA_WITH <- file.path('/home/imallona', 'cenp_chip', 'mapping_with_dupes')
DATA_WITHOUT <- file.path('/home/imallona', 'cenp_chip', 'mapping_without_dupes')

## if downsampled, only chr19
DOWNSAMPLE <- FALSE
```


# Ideogram with multimappers, on MACS2 peaks {.tabset .tabset-pills}

Definition of scores and signalValue: https://genome.ucsc.edu/FAQ/FAQformat.html#format12

```{r, results = 'asis', fig.width = 7, fig.height = 8}

fns <- list.files(DATA_WITH, pattern = "*narrowPeak", recursive = TRUE)

for (fn in fns) {
    cat('## ', fn, ' {.tabset .tabset-pills} \n\n')
    curr <- readPeakFile(file.path(DATA_WITH, fn))
    ## head(curr)
    ## ?covplot
    ## covplot(curr, weightCol="V5")
    cat('### score\n\n')
    print(covplot(curr, weightCol='V5'))
    cat('\n\n')
    
    cat('### signalValue\n\n')    
    print(covplot(curr, weightCol='V7'))
    cat('\n\n')

}
```

# Ideogram without multimappers, on MACS2 peaks {.tabset .tabset-pills}

Definition of scores and signalValue: https://genome.ucsc.edu/FAQ/FAQformat.html#format12

```{r, results = 'asis', fig.width = 7, fig.height = 8}

fns <- list.files(DATA_WITHOUT, pattern = "*narrowPeak", recursive = TRUE)

for (fn in fns) {
    cat('## ', fn, ' {.tabset .tabset-pills} \n\n')
    curr <- readPeakFile(file.path(DATA_WITHOUT, fn))

    cat('### score\n\n')
    print(covplot(curr, weightCol='V5'))
    cat('\n\n')
    
    cat('### signalValue\n\n')    
    print(covplot(curr, weightCol='V7'))
    cat('\n\n')
}
```

# Ideogram with multimappers, on BAM files  {.tabset .tabset-pills}

Reads a granges anyway...


```{bash}

# cd /home/imallona/cenp_chip/mapping_with_dupes/Bowtie2/

# conda activate snakePipes
# ## to get a modern samtools
# export PATH="/home/imallona/miniconda3//envs/ab0c3bd33e9b7239ee2b4f7d6bc5bca6/bin/:$PATH"

# for BAM in chip_HC_1.bam chip_PTZ_1.bam
# do
#     echo $BAM    
    
#     ## with a modern samtools
#     ## flags include also duplicates, secondary alignments etc
#     samtools depth -g 4091 $BAM > "$BAM".cov
#     grep -w chr19 "$BAM".cov > "$BAM".chr19.cov
#     gzip *cov
# done

# conda deactivate
```

```{bash, eval = FALSE}
cd /home/imallona/cenp_chip/mapping_with_dupes/Bowtie2/

mysql --user=genome --host=genome-mysql.cse.ucsc.edu -A -e \
      "select chrom, size from mm10.chromInfo" | grep -v size > mm10.genome

BEDTOOLS=~/soft/bedtools/bedtools-2.29.2/bin/bedtools
$BEDTOOLS makewindows -g mm10.genome -w 100000 | grep -w chr19 > chr19.windows.bed

for BAM in chip_HC_1.bam chip_PTZ_1.bam
do
    echo $BAM    

    $BEDTOOLS intersect -c -b $BAM -a chr19.windows.bed > "$BAM".bedtools.cov
done
```


```{r fig.width = 15, fig.height = 4}
DATA <- '/home/imallona/cenp_chip/mapping_with_dupes/Bowtie2'
fns <- list.files(DATA, pattern = ".*bam.bedtools.cov")
for (fn in fns) {
    cat('## ', fn, ' reads {.tabset .tabset-pills} \n\n')
    
    curr <- read.table(file.path(DATA, fn))
    head(curr)
    ## curr$end <- curr$V2 + 1000
    ## ?makeGRangesFromDataFrame

    curr$logreads <- log10(curr$V4 + 1)
    
    ## curr2 <-   makeGRangesFromDataFrame(curr,
    ##                                    seqnames.field= 'V1',
    ##                                    start.field="V2",
    ##                                    end.field="V3",
    ##                                    keep.extra.columns = TRUE)



    
    ## print(covplot(curr2, weightCol='V4', lower = 2))


    
    ## print(covplot(curr, weightCol='logreads'))
    ## plot(x = 1:nrow(curr), y = curr$V4, pch = 19)
    plot(x = 1:nrow(curr), y = curr$V4, pch = 19, xaxt = "n", main = 'chr19 coverage')
    axis(1, at = seq(1, nrow(curr), by = 100),
         labels =curr$V2[seq(1, nrow(curr), by = 100)] )
    

    cat('\n\n')

    cat('## ', fn, ' log reads {.tabset .tabset-pills} \n\n')
    
  
    
    ## print(covplot(curr, weightCol='logreads'))
    ## plot(x = 1:nrow(curr), y = curr$V4, pch = 19)
    plot(x = 1:nrow(curr), y = curr$logreads, pch = 19, xaxt = "n", main = 'chr19 coverage')
    axis(1, at = seq(1, nrow(curr), by = 100),
         labels =curr$V2[seq(1, nrow(curr), by = 100)] )
    
    rm(curr)

    cat('\n\n')
    
}
```

So clearly we masked the centromeres/pericentromeres


We'll map against the consensus centromeric reps, as in https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5567862/

Paired-end reads, resulting from either ChIP-seq or MNase-seq, were converted to a name-sorted SAM file using picard-tools and samtools [36], then joined in MATLAB using the ‘localalign’ function to determine the overlapping region between the paired-end reads (requiring ≥95% overlap identity; see [20]), and adapter sequences were removed if present. For analysis of minor and major satellite DNA, we used a custom tandem repeat analysis as described [20] with the following modifications. Joined reads were aligned to a trimerized mouse minor satellite consensus (GenBank: X14464.1) [37], to a dimerized mouse major satellite consensus (GenBank: V00846.1) [38], or to the reverse complement of those tandem consensus sequences. 




# Trace

```{r}
sessionInfo()
devtools::session_info()
```
